File: go/alphametics.go
package alphametics

import (
	"errors"
	"slices"
	"strings"
	"unicode"
)

func permutations(nums []int, k int) [][]int {
	result := [][]int{}
	visited := make([]bool, len(nums))
	current := []int{}
	backtrack(nums, visited, current, k, &result)
	return result
}

func backtrack(nums []int, visited []bool, current []int, k int, result *[][]int) {
	if len(current) == k {
		tmp := make([]int, k)
		copy(tmp, current)
		*result = append(*result, tmp)
		return
	}

	for i := 0; i < len(nums); i++ {
		if visited[i] {
			continue
		}
		visited[i] = true
		current = append(current, nums[i])
		backtrack(nums, visited, current, k, result)
		current = current[:len(current)-1]
		visited[i] = false
	}
}

func toNum(s string, mp map[string]int) (int, bool) {
	if len(s) > 1 && mp[string(s[0])] == 0 {
		return 0, false
	}
	n := 0
	for _, c := range s {
		n = n*10 + mp[string(c)]
	}
	return n, true
}

func check(vars []string, res string, mp map[string]int) bool {
	s := 0
	for _, v := range vars {
		n, ok := toNum(v, mp)
		if !ok {
			return false
		}
		s += n
	}

	r, ok := toNum(res, mp)
	if !ok {
		return false
	}

	return s == r
}

func Solve(puzzle string) (map[string]int, error) {
	mp := map[string]int{}
	for _, c := range puzzle {
		if unicode.IsLetter(c) {
			mp[string(c)] = 0
		}
	}

	keys := []string{}
	for k := range mp {
		keys = append(keys, k)
	}
	slices.Sort(keys)
	ps := permutations([]int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}, len(keys))

	arr := strings.Split(puzzle, " == ")
	rht := arr[len(arr)-1]
	vars := strings.Split(arr[0], " + ")

	for _, p := range ps {
		for i, k := range keys {
			mp[k] = p[i]
		}
		if check(vars, rht, mp) {
			return mp, nil
		}
	}

	return nil, errors.New("no solution")
}

File: elixir/alphametics.ex
defmodule Alphametics do
  @type puzzle :: binary
  @type solution :: %{required(?A..?Z) => 0..9}

  @doc """
  Takes an alphametics puzzle and returns a solution where every letter
  replaced by its number will make a valid equation. Returns `nil` when
  there is no valid solution to the given puzzle.

  ## Examples

    iex> Alphametics.solve("I + BB == ILL")
    %{?I => 1, ?B => 9, ?L => 0}

    iex> Alphametics.solve("A == B")
    nil
  """
  @spec solve(puzzle) :: solution | nil
  def solve(puzzle) do
    cs = puzzle 
      |> String.replace(~r/[^A-Z]/, "") 
      |> String.to_charlist()
      |> Enum.uniq()
      |> Enum.sort()

    [args, res] = String.split(puzzle, " == ")
    args = String.split(args, " + ")

    permutations(0..9, length(cs))
    |> Enum.reduce_while(nil, fn ns, acc ->
      mp = Enum.zip(cs, ns) |> Map.new()
      sum = Enum.reduce_while(args, 0, fn arg, sum ->
          case eval(arg, mp) do
            nil -> {:halt, nil}
            n -> {:cont, sum + n}
          end
        end)

      if sum != nil and sum == eval(res, mp) do
        {:halt, mp}
      else
        {:cont, acc}
      end
    end)
  end

  defp eval(cs, mp) do
    cs = String.to_charlist(cs)
    if mp[hd(cs)] == 0 do
      nil
    else
      cs |> Enum.reduce(0, fn c, acc -> mp[c] + acc * 10 end)
    end
  end
  
  defp permutations(lst, n, visited \\ %MapSet{}, curr \\ [], res \\ []) do
    if length(curr) == n do
      [curr|res]
    else
      lst
      |> Enum.reduce(res, fn i, acc ->
        if MapSet.member?(visited, i) do
          acc
        else
          permutations(lst, n, MapSet.put(visited, i), [i|curr], acc)
        end
      end)
    end
  end  
end

